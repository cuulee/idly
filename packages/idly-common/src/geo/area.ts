import area from '@turf/area';

export default area;
