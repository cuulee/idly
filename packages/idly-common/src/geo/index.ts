export * from '../geo/sphericalMercator';
export * from '../geo/bbox';
