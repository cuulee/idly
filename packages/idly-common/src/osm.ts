export type Entity = Node | Way | Relation;
export type Entities = Set<Node | Way | Relation>;
export type EntityId = string;
export type NodeId = string;
export type WayId = string;

export type ParentWays = Map<NodeId, Set<WayId>>;
// Table used to map id -> entity
export type EntityTable = Map<EntityId, Entity>;

export type Tags = Map<string, string>;

export type FeaturePropsTable = Map<EntityId, FeatureProps>;
export type FeatureProps = { [key: string]: string };

export type NodeGeometry =
  | OsmGeometry.POINT
  | OsmGeometry.VERTEX
  | OsmGeometry.VERTEX_SHARED;

export enum OsmGeometry {
  POINT = 'point',
  VERTEX = 'vertex',
  VERTEX_SHARED = 'vertex_shared', // the one shared among multiple ways
  AREA = 'area',
  LINE = 'line',
  RELATION = 'relation'
}

export enum EntityType {
  NODE = 'node',
  WAY = 'way',
  RELATION = 'relation'
}

export interface Node {
  readonly id: EntityId;
  readonly tags: Tags;
  readonly type: EntityType.NODE;
  readonly loc: LngLat;
  readonly attributes: Attributes;
}

export interface Way {
  readonly id: EntityId;
  readonly type: EntityType.WAY;
  readonly tags: Tags;
  readonly attributes: Attributes;
  readonly nodes: EntityId[];
}

export interface Relation {
  readonly id: EntityId;
  readonly type: EntityType.RELATION;
  readonly tags: Tags;
  readonly members: RelationMember[];
  readonly attributes: Attributes;
}

export interface Attributes {
  readonly visible?: boolean;
  readonly version?: number;
  readonly timestamp?: string;
  readonly changeset?: string;
  readonly uid?: string;
  readonly user?: string;
}

export interface RelationMember {
  readonly id: string;
  readonly type: string;
  readonly role: string;
}

export interface LngLat {
  readonly lat: number;
  readonly lon: number;
}
