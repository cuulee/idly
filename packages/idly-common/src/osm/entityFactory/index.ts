export { attributesFactory } from './attributesFactory';
export { nodeFactory } from './nodeFactory';
export { relationFactory } from './relationFactory';
export { tagsFactory } from './tagsFactory';
export { wayFactory } from './wayFactory';
