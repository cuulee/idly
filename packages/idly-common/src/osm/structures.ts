export type Entity = Node | Way | Relation;
export type Entities = Set<Node | Way | Relation>;
export type EntityId = string;
export type NodeId = string;
export type WayId = string;

export interface Tags {
  readonly [key: string]: string;
}

export type FeaturePropsTable = Map<EntityId, FeatureProps>;
export interface FeatureProps {
  [key: string]: string;
}

export type NodeGeometry = OsmGeometry.POINT | OsmGeometry.VERTEX;

export enum OsmGeometry {
  POINT = 'point',
  VERTEX = 'vertex',
  AREA = 'area',
  LINE = 'line',
  RELATION = 'relation',
}

export enum EntityType {
  NODE = 'node',
  WAY = 'way',
  RELATION = 'relation',
}

export interface Node {
  readonly id: EntityId;
  readonly tags: Tags;
  readonly type: EntityType.NODE;
  readonly loc: LngLat;
  readonly attributes: Attributes;
}

export interface Way {
  readonly id: EntityId;
  readonly type: EntityType.WAY;
  readonly tags: Tags;
  readonly attributes: Attributes;
  readonly nodes: ReadonlyArray<EntityId>;
}

export interface Relation {
  readonly id: EntityId;
  readonly type: EntityType.RELATION;
  readonly tags: Tags;
  readonly members: ReadonlyArray<RelationMember>;
  readonly attributes: Attributes;
}

export interface Attributes {
  readonly visible?: boolean;
  readonly version?: string;
  readonly timestamp?: string;
  readonly changeset?: string;
  readonly uid?: string;
  readonly user?: string;
}

// id & ref are the same thing
// id exists for backward compatibility with iD
// we are using ref
export interface RelationMember {
  readonly id: string;
  readonly ref: string;
  readonly type?: string;
  readonly role?: string;
}

export interface LngLat {
  readonly lat: number;
  readonly lon: number;
}
