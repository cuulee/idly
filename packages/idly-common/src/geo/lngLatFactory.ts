import { LngLat } from '../osm/structures';

export function lngLatFactory(
  obj: { lon: number; lat: number } | [number, number]
): LngLat {
  if (Array.isArray(obj)) {
    return { lon: obj[0], lat: obj[1] };
  }
  return { ...obj };
}
