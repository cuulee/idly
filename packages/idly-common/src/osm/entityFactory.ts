import { nodeFactory } from '../osm/nodeFactory';
import { relationFactory } from '../osm/relationFactory';
import { EntityType, Tags } from '../osm/structures';
import { Entity, Node, Relation, Way } from './structures';
import { tagsFactory } from './tagsFactory';
import { wayFactory } from './wayFactory';

export function entityFactory(entity: any): Entity {
  if (!entity.id) {
    throw new Error('not a valid entity');
  }

  if (typeof entity.tags === 'string') {
    entity.tags = tagsFactory(JSON.parse(entity.tags));
  } else if (Array.isArray(entity.tags)) {
    entity.tags = tagsFactory(entity.tags);
  } else if (!entity.hasOwnProperty('tags')) {
    entity.tags = tagsFactory([]);
  }

  if (!(entity.tags instanceof Tags)) {
    throw new Error('not proper tags');
  }

  switch (entity.type) {
    case EntityType.NODE:
      return nodeFactory(entity);

    case EntityType.WAY:
      return wayFactory(entity);

    case EntityType.RELATION:
      return relationFactory(entity);

    default:
      throw new Error('not a valid entity');
  }
}
