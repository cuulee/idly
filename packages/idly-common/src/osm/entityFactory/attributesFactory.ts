import { Attributes } from '../structures';

export function attributesFactory({
  visible,
  version,
  timestamp,
  changeset,
  uid,
  user,
}: {
  visible?: boolean;
  version?: string;
  timestamp?: string;
  changeset?: string;
  uid?: string;
  user?: string;
} = {}): Attributes {
  return {
    changeset,
    timestamp,
    uid,
    user,
    version,
    visible,
  };
}
