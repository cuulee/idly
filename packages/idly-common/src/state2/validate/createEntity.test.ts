import { lngLatFactory } from '../../geo/lngLatFactory';
import { attributesFactory } from '../../osm/entityFactory/attributesFactory';
import { Attributes } from '../../osm/structures';
import {
  createEntity,
  createNode,
  createRelation,
  createWay,
} from './createEntity';

const dummyAttributes = (): Attributes =>
  attributesFactory(
    Object.keys(attributesFactory())
      .map(r => [r, 'dummy-' + r])
      .reduce((prev: any, [k, v]) => {
        prev[k] = v;
        return prev;
      }, {})
  );

describe('createEntity', () => {
  it('should create', () => {
    expect(
      createEntity({
        attributes: dummyAttributes(),
        id: 'n-1',
        loc: lngLatFactory([0, 0]),
        tags: {},
        type: 'node',
      })
    ).toMatchSnapshot();
  });
  it('should work with strings', () => {
    expect(
      createEntity(
        JSON.stringify({
          attributes: dummyAttributes(),
          id: 'n-1',
          loc: lngLatFactory([0, 0]),
          tags: {},
          type: 'node',
        })
      )
    ).toBeTruthy();
  });
  it('should throw with wrong strings', () => {
    expect(() =>
      createEntity(JSON.stringify({}))
    ).toThrowErrorMatchingSnapshot();
  });
  it('should throw with empty object', () => {
    // @ts-ignore
    expect(() => createEntity({})).toThrowErrorMatchingSnapshot();
  });
  it('should throw when only id is provided', () => {
    // @ts-ignore
    expect(() => createEntity({ id: 'n-1' })).toThrowErrorMatchingSnapshot();
  });
  it('should throw when only id & tag is provided', () => {
    // @ts-ignore
    expect(() =>
      createEntity({ id: 'n-1', type: 'node' })
    ).toThrowErrorMatchingSnapshot();
  });
  it('should throw when no id', () => {
    expect(() =>
      // @ts-ignore
      createEntity({
        attributes: dummyAttributes(),
        loc: lngLatFactory([0, 0]),
        tags: {},
        type: 'node',
      })
    ).toThrowErrorMatchingSnapshot();
  });
  it('should throw when wrong id', () => {
    expect(() =>
      createEntity({
        attributes: dummyAttributes(),
        id: 'z-1',
        loc: lngLatFactory([0, 0]),
        tags: {},
        type: 'node',
      })
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      createEntity({
        attributes: dummyAttributes(),
        id: 'r-1',
        loc: lngLatFactory([0, 0]),
        tags: {},
        type: 'node',
      })
    ).toThrowErrorMatchingSnapshot();
  });
});

describe('createNode', () => {
  it('should create node', () => {
    expect(
      createNode({
        attributes: dummyAttributes(),
        id: 'n-1',
        loc: lngLatFactory([0, 0]),
        tags: {},
        type: 'node',
      })
    ).toMatchSnapshot();
  });
  it('should not create other entities', () => {
    expect(() =>
      createNode({
        attributes: dummyAttributes(),
        id: 'w-1',
        // @ts-ignore
        nodes: [],
        tags: {},
        type: 'way',
      })
    ).toThrowErrorMatchingSnapshot();
  });
});

describe('createWay', () => {
  it('should create way', () => {
    expect(
      createWay({
        attributes: dummyAttributes(),
        id: 'w-1',
        nodes: [],
        tags: {},
        type: 'way',
      })
    ).toMatchSnapshot();
  });
  it('should check for valid nodes', () => {
    expect(
      createWay({
        attributes: dummyAttributes(),
        id: 'w-1',
        nodes: ['n-1'],
        tags: {},
        type: 'way',
      })
    ).toMatchSnapshot();
  });
  it('should throw for invalid nodes', () => {
    expect(() =>
      createWay({
        attributes: dummyAttributes(),
        id: 'w-1',
        nodes: ['r-1'],
        tags: {},
        type: 'way',
      })
    ).toThrowErrorMatchingSnapshot();
  });
  it('should not create other entities', () => {
    expect(() =>
      createWay({
        attributes: dummyAttributes(),
        id: 'n-1',
        // @ts-ignore
        loc: lngLatFactory([0, 0]),
        tags: {},
        type: 'node',
      })
    ).toThrowErrorMatchingSnapshot();
  });
});

describe('createRelation', () => {
  it('should create relation', () => {
    expect(
      createRelation({
        attributes: dummyAttributes(),
        id: 'r-1',
        members: [
          {
            id: 'n-1',
          },
        ],
        tags: {},
        type: 'relation',
      })
    ).toMatchSnapshot();
  });
  it('should not create other entities', () => {
    expect(() =>
      createRelation({
        attributes: dummyAttributes(),
        id: 'w-1',
        // @ts-ignore
        nodes: [],
        tags: {},
        type: 'way',
      })
    ).toThrowErrorMatchingSnapshot();
  });
  it('should throw wrong members', () => {
    expect(() =>
      createRelation({
        attributes: dummyAttributes(),
        id: 'r-1',
        members: [
          {
            // @ts-ignore
            zebra: 'cross',
          },
        ] as any,
        tags: {},
        type: 'relation',
      })
    ).toThrowErrorMatchingSnapshot();
  });
});
