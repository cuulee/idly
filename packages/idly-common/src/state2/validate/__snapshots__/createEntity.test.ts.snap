// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createEntity should create 1`] = `
Object {
  "attributes": Object {
    "changeset": "dummy-changeset",
    "timestamp": "dummy-timestamp",
    "uid": "dummy-uid",
    "user": "dummy-user",
    "version": "dummy-version",
    "visible": "dummy-visible",
  },
  "id": "n-1",
  "loc": Object {
    "lat": 0,
    "lon": 0,
  },
  "tags": Object {},
  "type": "node",
}
`;

exports[`createEntity should throw when no id 1`] = `"Entity does not have a valid id"`;

exports[`createEntity should throw when only id & tag is provided 1`] = `"Invalid node"`;

exports[`createEntity should throw when only id is provided 1`] = `"Entity type is unknown"`;

exports[`createEntity should throw when wrong id 1`] = `"Invalid node"`;

exports[`createEntity should throw when wrong id 2`] = `"Invalid node"`;

exports[`createEntity should throw with empty object 1`] = `"Entity does not have a valid id"`;

exports[`createEntity should throw with wrong strings 1`] = `"Entity does not have a valid id"`;

exports[`createNode should create node 1`] = `
Object {
  "attributes": Object {
    "changeset": "dummy-changeset",
    "timestamp": "dummy-timestamp",
    "uid": "dummy-uid",
    "user": "dummy-user",
    "version": "dummy-version",
    "visible": "dummy-visible",
  },
  "id": "n-1",
  "loc": Object {
    "lat": 0,
    "lon": 0,
  },
  "tags": Object {},
  "type": "node",
}
`;

exports[`createNode should not create other entities 1`] = `"Entity must be node"`;

exports[`createRelation should create relation 1`] = `
Object {
  "attributes": Object {
    "changeset": "dummy-changeset",
    "timestamp": "dummy-timestamp",
    "uid": "dummy-uid",
    "user": "dummy-user",
    "version": "dummy-version",
    "visible": "dummy-visible",
  },
  "id": "r-1",
  "members": Array [
    Object {
      "id": "n-1",
    },
  ],
  "tags": Object {},
  "type": "relation",
}
`;

exports[`createRelation should not create other entities 1`] = `"Entity must be relation"`;

exports[`createRelation should throw wrong members 1`] = `"Invalid Relation"`;

exports[`createWay should check for valid nodes 1`] = `
Object {
  "attributes": Object {
    "changeset": "dummy-changeset",
    "timestamp": "dummy-timestamp",
    "uid": "dummy-uid",
    "user": "dummy-user",
    "version": "dummy-version",
    "visible": "dummy-visible",
  },
  "id": "w-1",
  "nodes": Array [
    "n-1",
  ],
  "tags": Object {},
  "type": "way",
}
`;

exports[`createWay should create way 1`] = `
Object {
  "attributes": Object {
    "changeset": "dummy-changeset",
    "timestamp": "dummy-timestamp",
    "uid": "dummy-uid",
    "user": "dummy-user",
    "version": "dummy-version",
    "visible": "dummy-visible",
  },
  "id": "w-1",
  "nodes": Array [],
  "tags": Object {},
  "type": "way",
}
`;

exports[`createWay should not create other entities 1`] = `"Entity must be way"`;

exports[`createWay should throw for invalid nodes 1`] = `"Invalid Way"`;
